src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.Unix.cs

```diff
@@ -944,6 +924,42 @@ private SortVersion GetSortVersion()
                                                              (byte) (LCID  & 0xFF)));
         }
 
+        private static class SortHandleCache
+        {
+            // in most scenarios there is a limited number of cultures with limited number of sort options
+            // so caching the sort handles and not freeing them is OK, see https://github.com/dotnet/coreclr/pull/25117 for more
+            private static readonly Dictionary<string, IntPtr> s_sortNameToSortHandleCache = new Dictionary<string, IntPtr>();
+
+            internal static IntPtr GetCachedSortHandle(string sortName)
+            {
+                lock (s_sortNameToSortHandleCache)
+                {
+                    if (!s_sortNameToSortHandleCache.TryGetValue(sortName, out IntPtr result))
+                    {
+                        var resultCode = Interop.Globalization.GetSortHandle(sortName, out result);
+
+                        if (resultCode == Interop.Globalization.ResultCode.OutOfMemory)
```

Comment:
Nit: The way this was written originally - check for success first and decide on the exception second - is a tiny bit more efficient. One vs. two conditions on the common path.

Reply:
In general, I agree that we should avoid having any additional branches on the hot path.

However, in this particular case, the method will be called once or twice per app lifetime (depending on the number of cultures and caching strategy `useUserOverride`) so having two branches is not an issue.

Adding one more `if` would increase the if nesting which imho would be worse.

```cs
lock (s_sortNameToSortHandleCache)
    if (!s_sortNameToSortHandleCache.TryGetValue(sortName, out IntPtr result))
        if (resultCode != resultCode.Success)
            if (resultCode == Interop.Globalization.ResultCode.OutOfMemory)
```

Reply:
I do think that the code would actually be more readable and easier to understand with the extra level of nesting.

Reply:
To me, the mapping between the ResultCode and Exception type belongs to ResultCode.

What do you say for the following:

```cs
// part of CompareInfo.Unix.cs
if (resultCode != Interop.Globalization.ResultCode.Success)
    throw resultCode.ToException();

// part of Interop.ResultCode.cs
internal static Exception ToException(this ResultCode resultCode)
{
    Debug.Assert(resultCode != ResultCode.Success);

    switch(resultCode)
    {
        case ResultCode.OutOfMemory:
            return new OutOfMemoryException();
        default:
            return new ExternalException(SR.Arg_ExternalException);
    }
}
```

In this particular use case, the ResultCode is mapped to Exception only once, but if we want to improve the code further this is my proposal.